runtime/riscv.S

```diff
@@ -0,0 +1,423 @@
+/*                                                                        */
+/*   Copyright 2016 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+/* Asm part of the runtime system, RISC-V processor, 64-bit mode */
+/* Must be preprocessed by cpp */
+
+#define ARG_DOMAIN_STATE_PTR t1
+#define DOMAIN_STATE_PTR s0
+#define TRAP_PTR s1
+#define ALLOC_PTR s10
+#define ALLOC_LIMIT s11
+#define TMP t0
+#define ARG t2
+
+#define STORE sd
+#define LOAD ld
+
+        .set    domain_curr_field, 0
+#define DOMAIN_STATE(c_type, name) \
+        .equ    domain_field_caml_##name, domain_curr_field ; \
+        .set    domain_curr_field, domain_curr_field + 1
+#include "../runtime/caml/domain_state.tbl"
+#undef DOMAIN_STATE
+
+#define Caml_state(var) (8*domain_field_caml_##var)(s0)
+
+#define FUNCTION(name) \
+        .align 2; \
+        .globl name; \
+        .type name, @function; \
+name:
+
+#if defined(__PIC__)
+        .option pic
+#define PLT(r) r@plt
+#else
+        .option nopic
+#define PLT(r) r
+#endif
+
+        .section        .text
+/* Invoke the garbage collector. */
+
+        .globl  caml_system__code_begin
+caml_system__code_begin:
+
+FUNCTION(caml_call_gc)
+.Lcaml_call_gc:
+        /* Record return address */
+        STORE   ra, Caml_state(last_return_address)
+        /* Record lowest stack address */
+        STORE   sp, Caml_state(bottom_of_stack)
+        /* Set up stack space, saving return address */
+        /* (1 reg for RA, 1 reg for FP, 21 allocatable int regs,
+            20 caller-save float regs) * 8 */
+        /* + 1 for alignment */
+        addi    sp, sp, -0x160
+        STORE   ra, 0x8(sp)
+        STORE   s0, 0x0(sp)
+        /* Save allocatable integer registers on the stack,
+           in the order given in proc.ml */
+        STORE   a0, 0x10(sp)
+        STORE   a1, 0x18(sp)
+        STORE   a2, 0x20(sp)
+        STORE   a3, 0x28(sp)
+        STORE   a4, 0x30(sp)
+        STORE   a5, 0x38(sp)
+        STORE   a6, 0x40(sp)
+        STORE   a7, 0x48(sp)
+        STORE   s2, 0x50(sp)
+        STORE   s3, 0x58(sp)
+        STORE   s4, 0x60(sp)
+        STORE   s5, 0x68(sp)
+        STORE   s6, 0x70(sp)
+        STORE   s7, 0x78(sp)
+        STORE   s8, 0x80(sp)
+        STORE   s9, 0x88(sp)
+        STORE   t2, 0x90(sp)
+        STORE   t3, 0x98(sp)
+        STORE   t4, 0xa0(sp)
+        STORE   t5, 0xa8(sp)
+        STORE   t6, 0xb0(sp)
```

Comment:
For RV-32 the GC expects an array of 32-bit integers, so the offsets must be spaced by 4 instead of by 8.